lib/events/athena/publisher.go

```diff
@@ -0,0 +1,162 @@
+	})
+
+	// TODO(tobiaszheller): consider reworking lib/observability to work also on s3 sdk-v2.
+	return &publisher{
+		topicARN: cfg.TopicARN,
+		snsPublisher: sns.NewFromConfig(*cfg.AWSConfig, func(o *sns.Options) {
+			o.Retryer = r
+		}),
+		uploader:      manager.NewUploader(s3.NewFromConfig(*cfg.AWSConfig)),
+		payloadBucket: cfg.largeEventsBucket,
+		payloadPrefix: cfg.largeEventsPrefix,
+	}
+}
+
+// EmitAuditEvent emits audit event to SNS topic. Topic should be connected with
+// SQS via subscription, so event is persisted on queue.
+// Events are marshaled as oneOf from apievents and encoded using base64.
+// For large events, payload is publihsed to S3, and on SNS there is only passed
+// location on S3.
+func (p *publisher) EmitAuditEvent(ctx context.Context, in apievents.AuditEvent) error {
+	// Just double check that audit event has minimum necessary fields for athena
+	// to works. Teleport emitter layer above makes sure that they are filled.
+	if in.GetID() == "" {
+		return trace.BadParameter("missing uid of audit event %s", in.GetType())
+	}
+	if in.GetTime().IsZero() {
+		return trace.BadParameter("missing time of audit event %s", in.GetType())
+	}
+
+	oneOf, err := apievents.ToOneOf(in)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	marshaledProto, err := oneOf.Marshal()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	b64Encoded := base64.StdEncoding.EncodeToString(marshaledProto)
+	if len(b64Encoded) > maxSNSMessageSize {
+		if len(b64Encoded) > maxS3BasedSize {
+			return trace.BadParameter("message too large to publish, size %d", len(b64Encoded))
+		}
+		return trace.Wrap(p.emitViaS3(ctx, in.GetID(), marshaledProto))
+	}
+	return trace.Wrap(p.emitViaSNS(ctx, in.GetID(), b64Encoded))
+}
+
+func (p *publisher) emitViaS3(ctx context.Context, uid string, marshaledEvent []byte) error {
+	path := filepath.Join(p.payloadPrefix, uid)
+	out, err := p.uploader.Upload(ctx, &s3.PutObjectInput{
+		Bucket: aws.String(p.payloadBucket),
+		Key:    aws.String(path),
+		Body:   bytes.NewBuffer(marshaledEvent),
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var versionID string
+	if out.VersionID != nil {
+		versionID = *out.VersionID
+	}
+	msg := &apievents.AthenaS3EventPayload{
+		Path:      path,
+		VersionId: versionID,
+	}
+	buf, err := msg.Marshal()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	_, err = p.snsPublisher.Publish(ctx, &sns.PublishInput{
+		TopicArn: aws.String(p.topicARN),
+		Message:  aws.String(string(buf)),
+		MessageAttributes: map[string]snsTypes.MessageAttributeValue{
+			payloadTypeAttr: {DataType: aws.String("String"), StringValue: aws.String(payloadTypeS3Based)},
+		},
+	})
+	return trace.Wrap(err)
+}
+
+func (p *publisher) emitViaSNS(ctx context.Context, uid string, b64Encoded string) error {
+	_, err := p.snsPublisher.Publish(ctx, &sns.PublishInput{
+		TopicArn: aws.String(p.topicARN),
+		Message:  aws.String(b64Encoded),
+		MessageAttributes: map[string]snsTypes.MessageAttributeValue{
```

Comment:
Don't these MessageAttributes get discarded when using `RawMessageDelivery = true`? I haven't tested this but am not sure: https://docs.aws.amazon.com/sns/latest/dg/sns-large-payload-raw-message-delivery.html. 

Only noticed since your example integration tests show it enabled.

Reply:
> When you enable raw message delivery for Amazon Kinesis Data Firehose or Amazon SQS endpoints, any Amazon SNS metadata is stripped from the published message and the message is sent as is.

For SQS, it strips only SNS metadata, `TopicArn` etc. MessageAttributes are preserved. I am using all time `RawMessageDelivery` and message attributes are present.